### PR TITLE
Ruby 3.0 Compatibility

### PR DESCRIPTION
**Pull request by pedro-stanaka**
I have seen the work that @cben put on #489 and felt like helping out with the last details. I hope this is good enough.

I think the most critical change is in `lib/kubeclient.rb` which I had to change the call to `Kubeclient::Client::apply_entity`.
Let me know if I can do something else.

Locally everything is working and tests are passing.

### CODE REVIEW

**Comment by cben** [comment]
Thanks, this is great help! :heart: 

Works for me locally as well but failed on github with unclear error about `rake`.  Any ideas?

**Comment by pedro-stanaka** [comment]
I will check this out. Seems something with bundler. I wanted to see the logs from github actions, but had to wait for your approval. I will take it from here.

**Comment by pedro-stanaka** [comment]
Oh man, I need approval on every push. @cben I fixed, it was some mistake from my side when changing the github actions config.

**Comment by pedro-stanaka** [comment]
Got ahold of this github actions thing. Will test on my repo and once I have final solution I will let you know here.

**Comment by pedro-stanaka** [comment]
So, there were two errors still. Small stuff. But we were affected by https://github.com/httprb/http/issues/572. I had to install rake in a previous step. 

Tested on my local and using github actions. Now it should be fine.

File: lib/kubeclient.rb [hunk 1]
```diff
@@ -322,7 +322,12 @@ def define_entity_methods
         end
 
         define_singleton_method("apply_#{entity.method_names[0]}") do |*args|
-          apply_entity(entity.resource_name, *args)
+          if RUBY_VERSION > '3.0'
+            rest = Hash[*args[1..-1]]
+            apply_entity(entity.resource_name, args[0], **rest)
+          else
+            apply_entity(entity.resource_name, *args)
+          end
```
[the remark this answers is not in the document]
Comment (cben) [reply lib/kubeclient.rb:330]: Not like that, fails on 3.0:
```
TestService#test_apply_service:
ArgumentError: wrong number of arguments (given 3, expected 2; required keyword: field_manager)
```
https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/#delegation
There is a way that works for all with `ruby2_keywords` but that itself is not defined before 2.7.
Let me try one more way first...

[the remark this answers is not in the document]
Comment (pedro-stanaka) [reply lib/kubeclient.rb:330]: Yep, that is exactly the problem I ran into. Could not figure out a better way to make it work on 3.0. I even asked for help for this one :rofl:

**Comment by pedro-stanaka** [comment]
Jeez! I think now I managed to fix the github action. Bundler + setup-ruby + bundler-cache simply didn't work for ruby 3.0. So for now I would say we give up cache and in return we get a green pipeline.

image

File: lib/kubeclient.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (cben) [reply lib/kubeclient.rb:330]: IIUC part of the problem for generic delegation is that `*args` gobbles the last hash in some ruby versions.
But we don't need generic, we know which positional args we take.  This version is closer in spirit to what we did for other methods:
```ruby
        define_singleton_method("apply_#{entity.method_names[0]}") \
        do |resource, opts = {}|
          apply_entity(entity.resource_name, resource, **opts)
        end
```

**Comment by cben** [comment]
BTW, for several other methods we didn't fully adopt the `**opts` style but stayed with `opts = {}`.
Why this works takes some time to figure out :face_with_head_bandage:, so for future reference:

> Note that Ruby 3.0 doesn’t behave differently when calling a method which doesn’t accept keyword arguments with keyword arguments. For instance, the following case is not going to be deprecated and will keep working in Ruby 3.0. The keyword arguments are still treated as a positional Hash argument.
> ```ruby
> def foo(kwargs = {})
>   kwargs
> end
>
> foo(k: 1) #=> {:k=>1}
> ```
> -- from https://www.ruby-lang.org/en/news/2019/12/12/separation-of-positional-and-keyword-arguments-in-ruby-3-0/#what-is-deprecated

So _receiving_ as positional is OK!
Now since some underlying methods like `get_entity` recieve as positional, it's also fine to pass it as positional from dynamic methods such as `get_#{entity.method_names[1]}`.

We can drop 2.4 and 2.5 already BTW, they are end-of-life, but supporting 2.6 means we can't do _much_ cleaner for now.

File: lib/kubeclient.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (pedro-stanaka) [reply lib/kubeclient.rb]: Thanks for this, yeah, you got it right. The *args will gobble the hash for options. And doing a double splat will error saying that it is a implicit conversion from array to hash. Thanks for this comment and for the other one, ruby is not my main language (for now) and this was really educative. Thanks for the effort in the review! :clap: 

I have applied this change now.

**Comment by pedro-stanaka** [comment]
> We can drop 2.4 and 2.5 already BTW, they are end-of-life, but supporting 2.6 means we can't do much cleaner for now.

I would leave this out of the scope of this MR, but I could contribute there as well, maybe do some tweaking on rubocop to see what we can gain from upgrading.

File: lib/kubeclient.rb (hunk 1 above)
[the remark this answers is not in the document]
Comment (pedro-stanaka) [reply lib/kubeclient.rb]: This change passed for all ruby versions! :raised_hands: https://github.com/pedro-stanaka/kubeclient/actions/runs/821447634

**Pull request merged**